### PR TITLE
Add option in llama.py to quantize weights to int8 at runtime

### PR DESCRIPTION
**Pull request by flammit**
Here's a summary of the memory and performance on METAL on my M2 Max with and without the `--quantize` option:

| Model Size | Loading Time | RAM | Fastest Ran Model Time |
| --- | --- | --- | --- | 
| 7B (fp16) | 1786.25 ms | 13.48GB | 105.30ms |
| 7B (int8) | 2608.36 ms | 6.87GB | 128.51 ms |
| 13B (fp16) | 6218.58 ms | 26.03 GB | 129.24 ms |
| 13B (int8) | 5576.42 ms | 13.19 GB | 169.25 ms |
| 30B (fp16) | 22970.66 ms | 65.06 GB | 1534.54 ms |
| 30B (int8) | 17363.83 ms | 32.75 GB | 258.79 ms | 
| 65B (fp16) | N/A | N/A | N/A |
| 65B (int8) | 40315.09 ms | 65.56 GB | 1551.57 ms | 

The performance of 30B (int8) is quite a bit better than 30B (fp16), perhaps due to memory bandwidth limitations.

### CODE REVIEW

**Comment by wozeparrot** [comment]
How does it perform on evals?

File: examples/llama.py [hunk 1]
```diff
@@ -130,11 +130,11 @@ def __call__(self, x:Tensor, start_pos:int, freqs_cis:Tensor, mask:Optional[Tens
...
 
 class Transformer:
-  def __init__(self, dim, multiple_of, n_heads, n_layers, norm_eps, vocab_size, max_batch_size=32, max_seq_len=1024):
-    self.layers = [TransformerBlock(dim, multiple_of, n_heads, norm_eps) for _ in range(n_layers)]
+  def __init__(self, dim, multiple_of, n_heads, n_layers, norm_eps, vocab_size, linear, max_batch_size=32, max_seq_len=1024):
+    self.layers = [TransformerBlock(dim, multiple_of, n_heads, norm_eps, linear) for _ in range(n_layers)]
     self.norm = RMSNorm(dim, norm_eps)
     self.tok_embeddings = Embedding(vocab_size, dim)
-    self.output = Linear(dim, vocab_size, bias=False)
+    self.output = AbsmaxQuantizedLinear(dim, vocab_size)
```
Comment (dc-dc-dc) [inline examples/llama.py:137]: should this also use the linear param passed?

Comment (flammit) [reply examples/llama.py:137]: Whoops - will fix that.

**Comment by flammit** [comment]
> How does it perform on evals?

I've only played with it in prompt mode to do sanity checking, but I'll see what it takes to add something to `model_eval.py` that's more rigorous.

**Comment by dc-dc-dc** [comment]
I believe the bounty references adding [lm-eval](https://github.com/EleutherAI/lm-evaluation-harness) testing alongside the int8 quantization

**Comment by prusnak** [comment]
I think it's worth merging PR https://github.com/tinygrad/tinygrad/pull/1284 first (it reduces code duplication for LLaMA), the rebasing this PR on top of that.

**Comment by flammit** [comment]
> I believe the bounty references adding [lm-eval](https://github.com/EleutherAI/lm-evaluation-harness) testing alongside the int8 quantization

Working on this now

**Comment by geohot** [comment]
Bounty is yours with the int8 + eval.

**Comment by flammit** [comment]
> Bounty is yours when you get the int8 + eval.
> 
> The 30B speed isn't memory bandwidth, it's amount of memory. You are probably swapping to disk for the FP16 one.

Cool thanks!

FWIW I'm on a 96GB machine so no swap for any of those runs.

**Comment by flammit** [comment]
Here are some results using `lm-eval` to execute the `arc_easy` tasks limited to 100 docs:

LLaMa 7B fp16:
|  Task  |Version| Metric |Value|   |Stderr|
|--------|------:|--------|----:|---|-----:|
|arc_easy|      0|acc     | 0.78|±  |0.0416|
|        |       |acc_norm| 0.71|±  |0.0456|

LLaMa 7B int8:
|  Task  |Version| Metric |Value|   |Stderr|
|--------|------:|--------|----:|---|-----:|
|arc_easy|      0|acc     | 0.75|±  |0.0435|
|        |       |acc_norm| 0.73|±  |0.0446|


LLaMa 13B fp16:
|  Task  |Version| Metric |Value|   |Stderr|
|--------|------:|--------|----:|---|-----:|
|arc_easy|      0|acc     | 0.77|±  |0.0423|
|        |       |acc_norm| 0.74|±  |0.0441|

LLaMa 13B int8:
|  Task  |Version| Metric |Value|   |Stderr|
|--------|------:|--------|----:|---|-----:|
|arc_easy|      0|acc     | 0.77|±  |0.0423|
|        |       |acc_norm| 0.73|±  |0.0446|

**Comment by dc-dc-dc** [comment]
can you merge the PR simplifying the scale param

**Comment by geohot** [comment]
Looking for the benchmarks from https://ai.meta.com/llama/

MMLU, TriviaQA, Natural Questions, GSM8K, HumanEval, AGIEval, BoolQ, HellaSwag, OpenBookQA, QuAC, Winogrande.

I'm fine with leaving a few out if they are annoying

**Comment by flammit** [comment]
> Looking for the benchmarks from https://ai.meta.com/llama/
> 
> MMLU, TriviaQA, Natural Questions, GSM8K, HumanEval, AGIEval, BoolQ, HellaSwag, OpenBookQA, QuAC, Winogrande.
> 
> I'm fine with leaving a few out if they are annoying

Seems like a bunch of these are supported now.  The full list is available here: https://github.com/EleutherAI/lm-evaluation-harness/blob/master/docs/task_table.md. 

Just need to specify which when running: i.e. ``--eval hellaswag --limit 10``.

File: examples/llama.py [hunk 2]
```diff
@@ -11,6 +11,10 @@
 np.set_printoptions(linewidth=200)
 from typing import Optional, Tuple
 
+from lm_eval.base import BaseLM
```
Comment (geohot) [inline examples/llama.py:14]: Shouldn't be required to run llama

**Comment by geohot** [comment]
The eval doesn't belong in llama.py. Can you make an external eval that runs as many of those as possible and post the results?

**Comment by flammit** [comment]
Here's the output from running the 6 benchmarks that are supported in `lm-eval`. I limited these to 100 tests each to have a reasonable run time.

7B-fp16:
```
python test/external/external_llama_eval.py --eval boolq,hellaswag,openbookqa,winogrande,triviaqa,gsm8k --limit 100 --size 7B
...
{
  "results": {
    "boolq": {
      "acc": 0.76,
      "acc_stderr": 0.04292346959909283
    },
    "hellaswag": {
      "acc": 0.57,
      "acc_stderr": 0.049756985195624284,
      "acc_norm": 0.79,
      "acc_norm_stderr": 0.040936018074033256
    },
    "openbookqa": {
      "acc": 0.37,
      "acc_stderr": 0.04852365870939099,
      "acc_norm": 0.46,
      "acc_norm_stderr": 0.05009082659620332
    },
    "winogrande": {
      "acc": 0.73,
      "acc_stderr": 0.04461960433384741
    },
    "triviaqa": {
      "em": 0.45,
      "em_stderr": 0.05
    },
    "gsm8k": {
      "acc": 0.0,
      "acc_stderr": 0.0
    }
  },
  "versions": {
    "boolq": 1,
    "hellaswag": 0,
    "openbookqa": 0,
    "winogrande": 0,
    "triviaqa": 2,
    "gsm8k": 0
  }
}
```

7B-int8:
```
python test/external/external_llama_eval.py --eval boolq,hellaswag,openbookqa,winogrande,triviaqa,gsm8k --limit 100 --size 7B --quantize
...
{
  "results": {
    "boolq": {
      "acc": 0.75,
      "acc_stderr": 0.04351941398892446
    },
    "hellaswag": {
      "acc": 0.58,
      "acc_stderr": 0.049604496374885836,
      "acc_norm": 0.77,
      "acc_norm_stderr": 0.04229525846816507
    },
    "openbookqa": {
      "acc": 0.36,
      "acc_stderr": 0.04824181513244218,
      "acc_norm": 0.47,
      "acc_norm_stderr": 0.05016135580465919
    },
    "winogrande": {
      "acc": 0.72,
      "acc_stderr": 0.04512608598542128
    },
    "triviaqa": {
      "em": 0.43,
      "em_stderr": 0.049756985195624284
    },
    "gsm8k": {
      "acc": 0.0,
      "acc_stderr": 0.0
    }
  },
  "versions": {
    "boolq": 1,
    "hellaswag": 0,
    "openbookqa": 0,
    "winogrande": 0,
    "triviaqa": 2,
    "gsm8k": 0
  }
}
```

File: examples/llama.py [hunk 3]
```diff
@@ -181,27 +170,86 @@ def sample(logits, temperature):
...
+      model = Transformer(**args_30B, linear=AbsmaxQuantizedLinear) if quantize else Transformer(**args_30B)
+      num_files = 4
+    elif model_size == "13B":
+      model = Transformer(**args_13B, linear=AbsmaxQuantizedLinear) if quantize else Transformer(**args_13B)
+      num_files = 2
+    else:
+      model = Transformer(**args_7B, linear=AbsmaxQuantizedLinear) if quantize else Transformer(**args_7B)
+      num_files = 1
+
+    weights = concat_weights([torch_load(filename) for filename in [f"{model_path}/{model_size}/consolidated.{i:02d}.pth" for i in range(num_files)]])
```
Comment (geohot) [inline examples/llama.py:222]: Want to confirm this is streaming...aka it quantizes the layers as it loads them. I want to avoid high peak memory usage.

Comment (flammit) [reply examples/llama.py:222]: Confirmed.  Memory increases linearly and peaks at the end once all the weights are loaded.

File: examples/llama.py [hunk 4]
```diff
@@ -51,8 +51,8 @@ def __call__(self, x:Tensor):
     return (x * (x.pow(2).mean(-1, keepdim=True) + self.eps).rsqrt()) * self.weight
 
 class Attention:
-  def __init__(self, dim, n_heads):
-    self.wq, self.wk, self.wv, self.wo = [Linear(dim, dim, bias=False) for _ in range(4)]
+  def __init__(self, dim, n_heads, linear=functools.partial(Linear, bias=False)):
```
[the remark this answers is not in the document]
Comment (flammit) [reply examples/llama.py]: Fixed in the latest commit.

**Comment by prusnak** [comment]
> Any additional comments or changes requested on this PR?

PR https://github.com/tinygrad/tinygrad/pull/1284 which removes the code reuse has been merged now. Please rebase on master

File: tinygrad/state.py [hunk 5]
```diff
@@ -72,7 +72,7 @@ def _rebuild_tensor_v2(storage, storage_offset, size, stride, requires_grad, bac
     # upstream LLaMA also does this conversion:
     # https://github.com/facebookresearch/llama/blob/6c7fe276574e78057f917549435a2554000a876d/llama/generation.py#L95
     if storage[1] == dtypes.bfloat16:
-      ret = ret.to("LLVM").half().to(Device.DEFAULT).realize()
+      ret = ret.to("LLVM").half().to(Device.DEFAULT)
```
Comment (flammit) [inline tinygrad/state.py:75]: Note @wozeparrot @prusnak - I removed this realize since it gets in the way of streaming the quanitzed data and spiking memory usage.

**Comment by flammit** [comment]
This PR has be rebased onto master and incorporated with the LLaMA-2 changes.   The support LLaMa-2 models and quantization have been tested to work and produce reasonable outputs.

File: tinygrad/state.py (hunk 5 above)
Comment (wozeparrot) [reply tinygrad/state.py:75]: should probably test with llama 2 to ensure that the cast is still happening, `.contiguous()` should enforce it a little bit better without requiring early realize.

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/state.py                          1               1
------------------------------------------------------------
lines added in the tinygrad folder: 0
```

**Comment by flammit** [comment]
Ran a quick comparison of LLaMA-2 7B on `arc_easy` for 100 docs:

| Model | acc | acc_norm |
| --- | --- | --- |
| 7B-fp16-v2 | 0.78 +/- 0.04 |  0.71 +/- 0.05 |
| 7B-int8-v2 |  0.75 +/- 0.04 | 0.73 +/- 0.04 |

**Comment by geohot** [comment]
LGTM

**Comment by prusnak** [comment]
I confirmed that LLaMA-2 quantized to 8-bits work on my M1 too. Thanks @flammit for significantly improving usability of LLaMA on CPUs! 👏
